### ISSUE TITLE
Training freezes when adding tf.function

### ISSUE THREAD

**Issue by netw0rkf10w** [tensorflow/tensorflow#33541]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): `v2.0.0-rc2-26-g64c3d38 2.0.0`
- Python version: 3.5.2
- CUDA/cuDNN version: 9.2
- GPU model and memory: GTX 1080 Ti

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

When adding the `tf.function` decorator to the `_epoch_train` (in the code below), it freezes.

**Code to reproduce the issue**
Code taken from this tutorial: https://www.tensorflow.org/tutorials/images/segmentation
Except that, instead of using `model.fit()`, I used the following custom class for the training loop:

```
import tensorflow as tf
class Trainer(object):
    def __init__(self, model, dataloaders, epochs=None,
                 steps=-1, log_interval=1):
        """
        dataloaders: {'train': train_loader, 'val': val_loader}
        Here we assume that the model has been compiled, i.e. it contains
            an optimizer and a loss function
        """
        self.model = model
        self.dataloaders = dataloaders
        self.epochs = epochs
        self.log_interval = log_interval
    
    def train(self):
        for epoch in range(self.epochs):
            message = '{}/{}:\t'.format(epoch, self.epochs)

            # Training phase
            self._epoch_train()
            # Tensorboard: https://www.tensorflow.org/tensorboard/get_started
            # Display metrics at the end of each epoch.
            for metric in self.model.metrics:
                metric_value = float(metric.result())
                message += 'train_{}: {}\t'.format(metric.name, metric_value)
                metric.reset_states()

            # Validation phase
            self._epoch_val()
            for metric in self.model.metrics:
                metric_value = float(metric.result())
                message += 'val_{}: {}\t'.format(metric.name, metric_value)
                metric.reset_states()

            print(message)

    @tf.function
    def _epoch_train(self):
        """
        Perform one training epoch
        """
        # Iterate over the batches of the dataset.
        # Note: use ds.enumerate() instead of enumerate(ds)
        # https://github.com/tensorflow/tensorflow/issues/30802
        for step, (x_batch, y_batch) in self.dataloaders['train'].enumerate():
            # Open a GradientTape to record the operations run
            # during the forward pass, which enables autodifferentiation.
            with tf.GradientTape() as tape:
                # Run the forward pass of the layer.
                # The operations that the layer applies
                # to its inputs are going to be recorded
                # on the GradientTape.
                logits = self.model(x_batch)  # Logits for this minibatch

                # Compute the loss value for this minibatch.
                loss_value = self.model.loss(y_batch, logits)

            # Use the gradient tape to automatically retrieve
            # the gradients of the trainable variables with respect to the loss.
            grads = tape.gradient(loss_value, self.model.trainable_weights)

            # Run one step of gradient descent by updating
            # the value of the variables to minimize the loss.
            self.model.optimizer.apply_gradients(zip(grads, self.model.trainable_weights))

            # Update metrics
            for metric in self.model.metrics:
                metric(y_batch, logits)

    def _epoch_val(self):
        """
        Perform one validation epoch
        """
        # Run a validation loop at the end of each epoch.
        for x_batch, y_batch in self.dataloaders['val']:
            logits = self.model(x_batch)
            # Update metrics
            for metric in self.model.metrics:
                metric(y_batch, logits)
```

**Comment by netw0rkf10w** [comment]
I guess the issue has something to do with iterating over the dataset, because if I take the training steps (together with `tf.function` out of the training epochs then it works:

    ```
@tf.function
    def _train_step(self, batch):
        x_batch, y_batch = batch
        with tf.GradientTape() as tape:
            # Run the forward pass of the layer.
            # The operations that the layer applies
            # to its inputs are going to be recorded
            # on the GradientTape.
            logits = self.model(x_batch)  # Logits for this minibatch

            # Compute the loss value for this minibatch.
            loss = self.model.loss(y_batch, logits)

        # Use the gradient tape to automatically retrieve
        # the gradients of the trainable variables with respect to the loss.
        grads = tape.gradient(loss, self.model.trainable_weights)

        # Run one step of gradient descent by updating
        # the value of the variables to minimize the loss.
        self.model.optimizer.apply_gradients(zip(grads, self.model.trainable_weights))

        # Update metrics
        for metric in self.model.metrics:
            metric(y_batch, logits)
        
        return loss

    @tf.function
    def _val_step(self, batch):
        x_batch, y_batch = batch
        logits = self.model(x_batch)
        for metric in self.model.metrics:
            metric(y_batch, logits)
```

**Comment by ravikyram** [comment]
@netw0rkf10w 

Looks like issue is resolved. Please, confirm.Thanks!

**Comment by netw0rkf10w** [comment]
@ravikyram The code in my first post is supposed to work, isn't it? Maybe this is a bug?

**Comment by ravikyram** [comment]
@netw0rkf10w 

Can you please help us with colab link  to reproduce the issue in our environment.It will be easy for localizing the issue faster.Thanks!

**Comment by ravikyram** [comment]
@netw0rkf10w 

Any updates please. Thanks!

**Comment by netw0rkf10w** [comment]
@ravikyram My apologies for the delay.
I tried to quickly reproduce the issue in [this notebook](https://colab.research.google.com/drive/11kDGNoNy2GbXMqpHEOTzCwM1-ZDR_28a), but failed (everything is working in the notebook).

I observed the issue when working on a segmentation task, so maybe I will take some time to see if it's reproducible on that dataset. For now let me close this issue.

Thanks again!
